File: 백준/Gold/3190. 뱀/뱀.js
// 뱀

const fs = require("fs");
const input = fs
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "./input.txt")
  .toString()
  .trim()
  .split("\n");

function sol() {
  const N = +input.shift();
  const K = +input.shift();
  const board = Array(N + 1)
    .fill()
    .map(() => Array(N + 1).fill(0));
  board[1][1] = 1;
  const apples = input.slice(0, K).map((e) => e.split(" ").map(Number));
  const L = +input.shift();
  const directionChange = {};
  input
    .slice(K)
    .map((e) => e.split(" "))
    .forEach((e) => (directionChange[+e[0]] = e[1]));

  const snake = [[1, 1]];
  const di = [0, 1, 0, -1];
  const dj = [1, 0, -1, 0];
  let direction = 0;
  let time = 0;

  for (const apple of apples) {
    board[apple[0]][apple[1]] = "a";
  }

  while (true) {
    time++;

    const head = snake.at(-1);
    let ni = head[0] + di[direction];
    let nj = head[1] + dj[direction];

    if (1 <= ni && ni <= N && 1 <= nj && nj <= N && board[ni][nj] !== 1) {
      snake.push([ni, nj]);

      if (board[ni][nj] !== "a") {
        let tail = snake.shift();
        board[tail[0]][tail[1]] = 0;
      }

      board[ni][nj] = 1;
    } else {
      return time;
    }

    if (directionChange[time]) {
      if (directionChange[time] == "D") {
        direction = (direction + 1) % 4;
      } else {
        direction = direction - 1 < 0 ? 4 - direction - 1 : direction - 1;
      }
    }
  }
}

console.log(sol());

File: 백준/Gold/3190. 뱀/뱀.py
from collections import deque
import sys

input = sys.stdin.readline

N = int(input())
K = int(input())

# 보드: 0 빈칸, 1 사과
board = [[0] * N for _ in range(N)]
for _ in range(K):
    r, c = map(int, input().split())
    board[r - 1][c - 1] = 1

L = int(input())
turns = {}
for _ in range(L):
    t, d = input().split()
    turns[int(t)] = d

# 방향: 오른쪽, 아래, 왼쪽, 위 (시계방향)
dx = [0, 1, 0, -1]
dy = [1, 0, -1, 0]
direction = 0  # 처음엔 오른쪽

snake = deque()
snake.append((0, 0))
snake_set = {(0, 0)}  # 몸 충돌 체크용

time = 0
x, y = 0, 0

while True:
    time += 1
    nx = x + dx[direction]
    ny = y + dy[direction]

    # 벽 충돌
    if not (0 <= nx < N and 0 <= ny < N):
        break

    # 자기 몸 충돌
    if (nx, ny) in snake_set:
        break

    # 머리 이동
    snake.append((nx, ny))
    snake_set.add((nx, ny))

    # 사과 있으면 꼬리 유지
    if board[nx][ny] == 1:
        board[nx][ny] = 0
    else:
        tx, ty = snake.popleft()
        snake_set.remove((tx, ty))

    # 방향 전환
    if time in turns:
        if turns[time] == 'L':
            direction = (direction - 1) % 4
        else:  # 'D'
            direction = (direction + 1) % 4

    x, y = nx, ny

print(time)